external/fv3fit/fv3fit/_shared/stacking.py

```diff
@@ -112,6 +112,7 @@ def shuffled(random: RandomState, datasets: Sequence[xr.Dataset]) -> xr.Dataset:
         datasets: input data to be shuffled, must contain identical
             dimensionality/coordinates if multiple datasets are given
     """
+    datasets = [dataset.unify_chunks() for dataset in datasets]
```

Comment:
Maybe I should add a logging message here if unifying chunks is necessary?

Reply:
This only makes sense to me if it's something we don't normally expect to happen. If we'd get this logging message every time we train, we'll just ignore it and add to console noise.

The docstring says we sample within chunks if chunking is present (i.e. without shuffling across chunks). Is this changed, and is there a reason for it to change?

Reply:
If I logged it, I would probably do it something like this:
```python
chunks_default = (len(datasets[0][SAMPLE_DIM_NAME]),)
try:
    chunks = datasets[0].chunks.get(SAMPLE_DIM_NAME, chunks_default)
except ValueError:
    datasets = [dataset.unify_chunks() for dataset in datasets]
    chunks = datasets[0].chunks.get(SAMPLE_DIM_NAME, chunks_default)
    logging.debug(f"Dataset variables had inconsistent chunk sizes; unified the chunks to: {chunks}")
```
So it would only show up if unification was required.  If `unify_chunks` changed the chunk size to something you didn't expect I think it would be useful to know.  Imagine if in an extreme example the chunk size along the sample dimension changed to `1` -- in that case no shuffling would occur.

This mainly comes up because `cos_zenith_angle` typically will not be chunked in time, since it is generally called with a time vector that is completely within memory.  `unify_chunks` is required to chunk it more finely along the sample dimension to match the other variables in the dataset.  The other variables should generally be chunked like each other -- if they weren't we would have encountered this error already.

But upon further reflection, I'm going to make this a moot point.  I think this issue should be solved upstream of fv3fit.  If we want `cos_zenith_angle` to be lazy in our training pipeline, we should find a way to compute it with the proper chunk structure from the start.  For now I have just modified the code in the DerivedMapping to always make its computation of `cos_zenith_angle` eager.  It was eager previously, and the performance improvement here is still a positive in that context.